src/pipx/standalone_python.py

```diff
@@ -0,0 +1,182 @@
+        shutil.rmtree(download_dir, ignore_errors=True)
+
+    return str(install_dir / "bin/python3")
+
+
+def get_or_update_index():
+    """Get or update the index of available python builds from
+    the python-build-standalone repository."""
+    index_file = PIPX_STANDALONE_PYTHON_CACHEDIR / "index.json"
+    if index_file.exists():
+        index = json.loads(index_file.read_text())
+        # update index if it's been more than a month
+        fetched = datetime.datetime.fromtimestamp(index["fetched"])
+        if datetime.datetime.now() - fetched > datetime.timedelta(days=30):
+            index = {}
+    else:
+        index = {}
+    if not index:
+        releases = get_latest_python_releases()
+        index = {"fetched": datetime.datetime.now().timestamp(), "releases": releases}
+        # update index
+        index_file.write_text(json.dumps(index))
+    return index
+
+
+def get_latest_python_releases() -> List[str]:
+    """Returns the list of python download links from the latest github release."""
+    try:
+        with urlopen(GITHUB_API_URL) as response:
+            release_data = json.load(response)
+
+    except urllib.error.URLError as e:
+        # raise
+        raise PipxError("Unable to fetch python-build-standalone release data.") from e
+
+    return [asset["browser_download_url"] for asset in release_data["assets"]]
+
+
+def list_pythons() -> Dict[str, str]:
+    """Returns available python versions for your machine and their download links."""
+    system, machine = platform.system(), platform.machine()
+    download_link_suffix = MACHINE_SUFFIX[system][machine]
+    # linux suffixes are nested under glibc or musl builds
+    if system == "Linux":
+        # fallback to musl if libc version is not found
+        libc_version = platform.libc_ver()[0] or "musl"
+        download_link_suffix = download_link_suffix[libc_version]
+
+    python_releases = get_or_update_index()["releases"]
+
+    available_python_links = [link for link in python_releases if link.endswith(download_link_suffix)]
+
+    python_versions: dict[str, str] = {}
+    for link in available_python_links:
+        match = PYTHON_VERSION_REGEX.search(link)
+        assert match is not None
+        python_version = match[1]
+        python_versions[python_version] = link
+
+    sorted_python_versions = {
+        version: python_versions[version]
+        for version in sorted(
+            python_versions,
+            # sort by semver
+            key=lambda version: [int(k) for k in version.split(".")],
+            reverse=True,
+        )
+    }
+    return sorted_python_versions
+
+
+def resolve_python_version(requested_version: str):
+    pythons = list_pythons()
+
+    for version, version_download_link in pythons.items():
+        if version.startswith(requested_version):
+            python_version = version
+            download_link = version_download_link
+            break
+    else:
```

Comment:
This seems to be indented wrongly?

Reply:
That is a for-else loop. if the for loop is exhausted without breaking (without finding a valid python version), the else clause is executed and raises an exception.